### ISSUE TITLE
Force Atlas 2 

### ISSUE THREAD

**Issue by vivekdaryani** [jacomyal/sigma.js#686]
Can we apply ForceAtlas2 on some nodes of graph ( not on complete graph ) ????

**Comment by AVermeij** [comment]
Sure, you just have to tell the supervisor to filter the nodes and edges according to which nodes you want to apply the layout to. For instance, this only runs FA2 on non-hidden nodes:

``
  Supervisor.prototype.applyLayoutChanges = function(initial) {
    var nodes = this.graph.nodes().filter(function(n) {return !n.hidden})
``

and 

``
  Supervisor.prototype.graphToByteArrays = function() {
    var nodes = this.graph.nodes().filter(function(n) {return !n.hidden}),
        edges = this.graph.edges().filter(function(e) {return !e.hidden}),
``

**Comment by vivekdaryani** [comment]
From my all nodes (non-hidden), i want to apply the force layout on some nodes according to its type
for example, apply layout to those nodes only which has node type "square"

**Comment by AVermeij** [comment]
This should do that (simply replace in the above):

``
this.graph.nodes().filter(function(n) {return n.type === 'square'})
``

**Comment by Compynerd255** [comment]
I would assume that the code to follow each of these examples is the body of the old prototype, or a call to it?

**Comment by AVermeij** [comment]
I'm not sure whether I fully understand your question, but the code which I'm referring to is this file: 

https://github.com/jacomyal/sigma.js/blob/master/plugins/sigma.layout.forceAtlas2/supervisor.js

in lines 121/122 and 165

**Comment by Compynerd255** [comment]
@AVermeij That's what I meant - if you were referring to modifying or reprototyping the supervisor code. I guess that's a good solution for now, though it would seem better if we could assign a `fixed` boolean on the node in order to keep it from moving, so that Force Atlas 2 doesn't run calculations on it. You could also get away with adding a mass feature - an extremely large (or perhaps infinite) `mass` on a node would keep it from moving very much.

**Comment by vivekdaryani** [comment]
@ Compynerd255 ForceAtlas2 run calculations on that node which has a "fixed :  true"

**Comment by vivekdaryani** [comment]
@AVermeij  the above changes in lines 121/122 and 165 are not working.

**Comment by Compynerd255** [comment]
@vivekdaryani No, that feature has not been implemented yet. I'm just saying that it would be really nice if it was.

**Comment by AVermeij** [comment]
@vivekdaryani Can you share your code, including how you set your ``n.type``? Because it works fine here.

**Comment by vivekdaryani** [comment]
@AVermeij  Your solution works, thanks

**Comment by Yomguithereal** [comment]
The thing that must be taken into account here is if you want the non-moving nodes to alter the layout or not (because they do have an existence and alter the layout's gravity).

**Comment by vivekdaryani** [comment]
@Yomguithereal Is there any way that moving nodes do not overlap the non-moving nodes??

**Comment by andreagobbi** [comment]
@Yomguithereal adding the lines suggested by @AVermeij the  hidden nodes alter the gravity of the layout. Any suggestion for remove the effects of hidden nodes? I'll try by my own but if someone already have a solution please share it :-)
THX

**Comment by andreagobbi** [comment]
I have found this solution, in function
`  Supervisor.prototype.graphToByteArrays 
`
I added these lines
```
  var val=this.graph.degree(nodes[i].id)
      if(!(typeof nodes[i].true_degree ==='undefined'))
        val=nodes[i].true_degree
      // Populating byte array
...
      this.nodesByteArray[j + 6] = 1 + val;
...
```

And then `true_degree ` can be easlilly update in my original script.

**Comment by andreagobbi** [comment]
I suggest also to use the @AVermeij suggestion in the following files:

> sigma.classes.camera.js, sigma.core.js, sigma.middlewares.rescale

in order to speedup the Force Atlas 2 layout engine. This because the `refresh` function takes long time also if few nodes and edges are displayed (in my case the origninal graph has 60k nodes and 200k edges, initially it is empty then the user can enlarge the graph searching a node or clicking a neighboor ...)

**Comment by efleury** [comment]
I did update supervisor.js with the filter line, but if forceatlas is working on the global graph, when selecting only few nodes and edges (on this point it seems that puting hidden to true to a node does not put all adjacent edges hidden...)  it appears  that afeter 2 iterations in the method applyLayoutChanges  the position of the nodes become NaN and the result is that the graph disappear.... any clue ?

**Comment by apitts** [comment]
@efleury I recall encountering something similar when I neglected to filter some edges that no longer had a source or target after filtering some nodes. Perhaps check if that is your issue.
